### PR TITLE
IPC4: Add support for HDA/DMIC/SSP DAI's

### PR DESCRIPTION
**Pull request by ranj063**
This PR completes the support for HDA/DMIC/SSP DAI support with ipc4.

### CODE REVIEW

File: sound/soc/sof/intel/hda-dai.c [hunk 1]
```diff
@@ -340,6 +344,88 @@ static int ipc3_hda_link_pcm_trigger(struct snd_pcm_substream *substream,
...
+		break;
+	case SNDRV_PCM_TRIGGER_SUSPEND:
+	case SNDRV_PCM_TRIGGER_STOP:
+	{
+		struct snd_sof_widget *pipe_widget = swidget->pipe_widget;
+		struct sof_ipc4_pipeline *pipeline = pipe_widget->private;
+
+		ret = sof_ipc4_set_pipeline_state(sdev, swidget->pipeline_id,
+						  SOF_IPC4_PIPE_PAUSED);
+		if (ret < 0)
```
Comment (RanderWang) [inline sound/soc/sof/intel/hda-dai.c:487]: we can set pipeline state in pcm_trigger or dai_trigger. why is here ? thanks

File: sound/soc/sof/ipc4-topology.c [hunk 2]
```diff
@@ -1315,6 +1315,47 @@ static int sof_ipc4_dai_config(struct snd_sof_dev *sdev, struct snd_sof_widget *
...
+	u32 len_check;
+	int i;
+
+	manifest = (struct sof_manifest *)man_ptr;
+	manifest_tlv = manifest->items;
+	len_check = sizeof(struct sof_manifest);
+	for (i = 0; i < manifest->count; i++) {
+		len_check += sizeof(struct sof_manifest_tlv) + manifest_tlv->size;
+		if (len_check > size)
+			return -EINVAL;
```
Comment (RanderWang) [inline sound/soc/sof/ipc4-topology.c:1480]: I prefer a white space line here

File: sound/soc/sof/ipc4-topology.c [hunk 3]
```diff
@@ -819,6 +837,133 @@ static void sof_ipc4_unprepare_copier_module(struct snd_sof_widget *swidget)
...
+				     int *sample_rate, int *channel_count, int *bit_depth)
+{
+	int depth;
+
+	switch (params_format(params)) {
+	case SNDRV_PCM_FORMAT_S16_LE:
+		depth = 16;
+		break;
+	case SNDRV_PCM_FORMAT_S24_LE:
+		depth = 24;
```
Comment (RanderWang) [inline sound/soc/sof/ipc4-topology.c]: is depth for container bits size or valid sample bits ?

File: sound/soc/sof/ipc4-topology.c [hunk 4]
```diff
@@ -819,6 +837,133 @@ static void sof_ipc4_unprepare_copier_module(struct snd_sof_widget *swidget)
...
+		return ret;
+
+	dev_dbg(sdev->dev, "%s: dai_index %d nhlt_type %d dir %d\n",
+		__func__, dai_index, nhlt_type, dir);
+
+	/* try to find matching blob against hw_config params */
+	cfg = intel_nhlt_get_endpoint_blob(sdev->dev, ipc4_data->nhlt, dai_index, nhlt_type,
+					   bit_depth, bit_depth, channel_count, sample_rate,
+					   dir, 0);
+
```
Comment (RanderWang) [inline sound/soc/sof/ipc4-topology.c:926]: 10 parameters,  not a good programming practice

File: sound/soc/sof/ipc4-topology.c [hunk 5]
```diff
@@ -819,6 +837,133 @@ static void sof_ipc4_unprepare_copier_module(struct snd_sof_widget *swidget)
...
+		ret = snd_sof_get_pcm_hw_params(sdev, params, &sample_rate, &channel_count,
+						&bit_depth);
+		break;
+	case SOF_DAI_INTEL_SSP:
+		nhlt_type = NHLT_LINK_SSP;
+		ret = snd_sof_get_hw_config_params(sdev, dai, &sample_rate, &channel_count,
+						   &bit_depth);
+		break;
+	default:
+		return 0;
```
Comment (RanderWang) [inline sound/soc/sof/ipc4-topology.c:913]: do we need to set len & dst here ?

File: sound/soc/sof/ipc4-topology.h [hunk 6]
```diff
@@ -31,6 +31,12 @@
 #define SOF_IPC4_NODE_INDEX(x)	((x) & SOF_IPC4_NODE_INDEX_MASK)
 #define SOF_IPC4_NODE_TYPE(x)  ((x) << 8)
 
+/* Node ID for SSP type DAI copiers */
+#define SOF_IPC4_NODE_INDEX_INTEL_SSP(x) (((x) & 0xf) << 4)
```
Comment (RanderWang) [inline sound/soc/sof/ipc4-topology.h:35]: bit 7 ~ 4 is for instance id like our dai id. Can we define a macro #define xxx_INSTANCE_OFFSET 4 (choose what you like, I am not good at naming).

File: sound/soc/sof/ipc4-topology.c (hunk 3 above)
Comment (bardliao) [reply sound/soc/sof/ipc4-topology.c]: > is depth for container bits size or valid sample bits ?

Both

File: sound/soc/sof/intel/hda-dai.c (hunk 1 above)
Comment (ujfalusi) [reply sound/soc/sof/intel/hda-dai.c:487]: I think we need to sandwich the DSP and host operations and this is why there is a mix of sof_ipc4_* and host ops here.

I just wonder how this plugs in with the pcm ops in core level..

File: sound/soc/sof/intel/hda-dai.c [hunk 7]
```diff
@@ -516,6 +618,59 @@ static const struct snd_soc_dai_ops ipc3_ssp_dai_ops = {
...
+			return ret;
+		pipeline->state = SOF_IPC4_PIPE_RESET;
+		break;
+	case SNDRV_PCM_TRIGGER_PAUSE_PUSH:
+		ret = sof_ipc4_set_pipeline_state(sdev, swidget->pipeline_id,
+						  SOF_IPC4_PIPE_PAUSED);
+		if (ret < 0)
+			return ret;
+		pipeline->state = SOF_IPC4_PIPE_PAUSED;
```
Comment (ujfalusi) [inline sound/soc/sof/intel/hda-dai.c:745]: The code in `ipc4-pcm.c` in core also does `sof_ipc4_set_pipeline_state()` in trigger, how these place are not going to race and conflict?

File: include/uapi/sound/sof/abi.h [hunk 8]
```diff
@@ -59,4 +61,34 @@
...
+
+/**
+ * struct sof_manifest_tlv - SOF manifest TLV data
+ * @type: type of data
+ * @size: data size (not including the size of this struct)
+ * @data: payload data
+ */
+struct sof_manifest_tlv {
+	__u32 type;
+	__u32 size;
```
Comment (ujfalusi) [inline include/uapi/sound/sof/abi.h]: Does the endianness matters here?
Where this data is coming from?
Is `abi.h` is the right place for this?

fwiw, include/sound/soc/sof/ipc4/header.h has similar sog_manifest_tlv struct, but it is named as sof_ipc4_tuple and I'm sure that the same thing runs under different name elsewhere...

File: sound/soc/sof/ipc4-topology.c [hunk 9]
```diff
@@ -819,6 +837,133 @@ static void sof_ipc4_unprepare_copier_module(struct snd_sof_widget *swidget)
...
+	case SNDRV_PCM_FORMAT_S24_LE:
+		depth = 24;
+		break;
+	case SNDRV_PCM_FORMAT_S32_LE:
+		depth = 32;
+		break;
+	default:
+		dev_err(sdev->dev, "invalid pcm frame format %d\n", params_format(params));
+		return -EINVAL;
+	}
```
Comment (ujfalusi) [inline sound/soc/sof/ipc4-topology.c]: depth = params_width(params);

File: sound/soc/sof/ipc4-topology.c [hunk 10]
```diff
@@ -819,6 +837,133 @@ static void sof_ipc4_unprepare_copier_module(struct snd_sof_widget *swidget)
...
+
+	*bit_depth = depth;
+	*channel_count = params_channels(params);
+	*sample_rate = params_rate(params);
+
+	dev_dbg(sdev->dev, "%s: rate %d channels %d sample_valid_bytes %d\n",
```
Comment (ujfalusi) [inline sound/soc/sof/ipc4-topology.c]: sample_valid_bytes is in bits...

I would use different order: rate, sample_width, channels
and probably in a form of 48000Hz, 24bits, 2 channels (<- but it is likely only me).

File: sound/soc/sof/ipc4-topology.c (hunk 5 above)
Comment (ujfalusi) [reply sound/soc/sof/ipc4-topology.c:913]: is it fine to ignore unknown linktype?

File: sound/soc/sof/ipc4-topology.c (hunk 4 above)
Comment (ujfalusi) [reply sound/soc/sof/ipc4-topology.c:926]: it is existing code external code.

File: sound/soc/sof/ipc4-topology.c [hunk 11]
```diff
@@ -819,6 +837,133 @@ static void sof_ipc4_unprepare_copier_module(struct snd_sof_widget *swidget)
...
+		__func__, dai_index, nhlt_type, dir);
+
+	/* try to find matching blob against hw_config params */
+	cfg = intel_nhlt_get_endpoint_blob(sdev->dev, ipc4_data->nhlt, dai_index, nhlt_type,
+					   bit_depth, bit_depth, channel_count, sample_rate,
+					   dir, 0);
+
+	if (!cfg) {
+		dev_err(sdev->dev,
+			"no matching blob for bit_depth: %d channels: %d sample_rate: %d\n",
```
Comment (ujfalusi) [inline sound/soc/sof/ipc4-topology.c]: Would be nice to have some consistence in how we print.
line 864: "%s: rate %d channels %d sample_valid_bytes %d\n"
here:       "bit_depth: %d channels: %d sample_rate: %d\n"

4800Hz, 24bits, 2 channels as a format to print in all places? Or something, but consistent?

File: sound/soc/sof/ipc4-topology.h (hunk 6 above)
Comment (ujfalusi) [reply sound/soc/sof/ipc4-topology.h:35]: I would not bother with another define which is only used in this macro.

File: sound/soc/sof/ipc4-topology.c [hunk 12]
```diff
@@ -1508,6 +1508,63 @@ static int sof_ipc4_parse_manifest(struct snd_soc_component *scomp, int index,
...
+	case SOF_DAI_INTEL_SSP:
+		switch (clk_type) {
+		case SOF_DAI_CLK_INTEL_SSP_MCLK:
+			return le32_to_cpu(hw_config->mclk_rate);
+		case SOF_DAI_CLK_INTEL_SSP_BCLK:
+			return le32_to_cpu(hw_config->bclk_rate);
+		default:
+			break;
+		}
+		dev_err(sdev->dev, "failed to get SSP clk %d rate\n", clk_type);
```
Comment (ujfalusi) [inline sound/soc/sof/ipc4-topology.c]: move this within the clk_type case's default?

Comment (ujfalusi) [reply sound/soc/sof/ipc4-topology.c]: and re-phrase the error.

File: sound/soc/sof/core.c [hunk 13]
```diff
@@ -360,6 +360,13 @@ int snd_sof_device_probe(struct device *dev, struct snd_sof_pdata *plat_data)
 	sdev->first_boot = true;
 	dev_set_drvdata(dev, sdev);
 
+	/* check IPC support */
+	if (!(BIT(plat_data->ipc_type) & plat_data->desc->ipc_supported_mask)) {
+		dev_err(dev, "ipc_type %d is not supported on this platform, mask is %#x\n",
+			plat_data->ipc_type, plat_data->desc->ipc_supported_mask);
+		return -EINVAL;
+	}
```
Comment (ujfalusi) [inline sound/soc/sof/core.c:367]: this patch is kind of an oddball within the series, can it be a separate PR?

**Review by ujfalusi — CHANGES_REQUESTED** [review]

File: sound/soc/sof/intel/hda-dai.c (hunk 1 above)
Comment (ranj063) [reply sound/soc/sof/intel/hda-dai.c:487]: @RanderWang for HDA DAI's we should pause the pipeline, clear the run BIT for the link DMA and then reset it. How can we clear the link DMA run BIT in the pcm trigger?

File: include/uapi/sound/sof/abi.h (hunk 8 above)
Comment (ranj063) [reply include/uapi/sound/sof/abi.h]: @ujfalusi the data is coming from the topology binary and I'd guess that the endianness does matter. I'm not sure abi.h is the right place for it but then it is for FW ABI-specific binary data so I guess its ok, but not sure. @juimonen can you please comment?

File: sound/soc/sof/ipc4-topology.c (hunk 5 above)
Comment (ranj063) [reply sound/soc/sof/ipc4-topology.c:913]: no, we don't need to set len and dst here and yes, we should ignore the other DAI types because they do not use NHLT blobs.

File: sound/soc/sof/ipc4-topology.h (hunk 6 above)
Comment (ranj063) [reply sound/soc/sof/ipc4-topology.h:35]: right only ever used or called once.

File: sound/soc/sof/core.c (hunk 13 above)
Comment (ranj063) [reply sound/soc/sof/core.c:367]: right but without this, you can't test this PR as is with IPC4.

File: sound/soc/sof/intel/hda-dai.c (hunk 7 above)
Comment (ranj063) [reply sound/soc/sof/intel/hda-dai.c:745]: @ujfalusi they cannot race, this is done during BE DAI trigger and the code in ipc4-pcm.c is invoked during FE DAI trigger. They are already protected by the PCM mutex in the core isnt it?

File: sound/soc/sof/ipc4-topology.c (hunk 2 above)
Comment (ranj063) [reply sound/soc/sof/ipc4-topology.c:1480]: wish granted!

File: include/uapi/sound/sof/abi.h (hunk 8 above)
Comment (juimonen) [reply include/uapi/sound/sof/abi.h]: is there also a "super" struct holding these ipc4_tuples? I mean I can use ipc4_tuple but I still need the master struct describing the amount of items etc. or this is just about the location/file of the structs?

Comment (juimonen) [reply include/uapi/sound/sof/abi.h]: @ujfalusi so the tlv is in abi.h as the original sof_manifest was in abi.h, I just "augmented it". You're right that the tlv struct is similar/same than ipc4_tuple, but it would just look strange if I would use it as generic manifest data tlv. If there's somewhere else in alsa a generic similar struct to use, that would perhaps make more sense. People would just start to wonder what this got to do with ipc4...

File: sound/soc/sof/intel/hda-dai.c (hunk 1 above)
Comment (RanderWang) [reply sound/soc/sof/intel/hda-dai.c:487]: so, it is only for HDA DAI ?

File: sound/soc/sof/ipc4-topology.h (hunk 6 above)
Comment (RanderWang) [reply sound/soc/sof/ipc4-topology.h:35]: for bit 0 ~ 3, we don't use it now. it is for virtual slot. Maybe it will be used in future. We can add macro at that time

File: sound/soc/sof/intel/hda-dai.c (hunk 7 above)
Comment (RanderWang) [reply sound/soc/sof/intel/hda-dai.c:745]: is BE DAI triggered before FE DAI ?  As I know it depends on some flags. According to above answer we need to set state in BE DAI, so BE DAI should be triggered before FE DAI

Comment (ranj063) [reply sound/soc/sof/intel/hda-dai.c:745]: @RanderWang yes, I take care of it with the be_dai_link_fixup() in ipc4-pcm.c

File: sound/soc/sof/intel/hda-dai.c (hunk 1 above)
Comment (ranj063) [reply sound/soc/sof/intel/hda-dai.c:487]: it is only really required for HDA DAI but the reason why we add the BE DAI trigger for all types of DAI's is because in the FE DAI trigger, I do not know what my BE DAI type is in order to not trigger the BE pipelines

File: include/uapi/sound/sof/abi.h (hunk 8 above)
Comment (ujfalusi) [reply include/uapi/sound/sof/abi.h]: @juimonen, there is `struct snd_ctl_tlv` in include/uapi/sound/asound.h, but it is not **packed**, which is needed for us when communicating with the firmware or reading file. It is using `unsigned int`, which is u32 mostly, but for inter-processor communication we should be strict on type and size.
But the concept is not that innovative, I'm sure it is used in lot of places (does IPC3 uses similar construct?). Anyways, if you have a good generic name for it, we can switch to use it in all places in SOF.

File: include/uapi/sound/sof/abi.h [hunk 14]
```diff
@@ -59,4 +61,34 @@
...
+ * @abi_major: Major ABI version
+ * @abi_minor: Minor ABI version
+ * @abi_patch: ABI patch
+ * @count: count of tlv items
+ * @items: consecutive variable size tlv items
+ */
+struct sof_manifest {
+	__u8 abi_major;
+	__u8 abi_minor;
+	__u8 abi_patch;
```
Comment (ujfalusi) [inline include/uapi/sound/sof/abi.h]: fyi: The firmware uses u16 fort he version numbers.

File: sound/soc/sof/core.c (hunk 13 above)
Comment (ujfalusi) [reply sound/soc/sof/core.c:367]: OK

File: sound/soc/sof/ipc4-topology.c [hunk 15]
```diff
@@ -1315,6 +1438,105 @@ static int sof_ipc4_dai_config(struct snd_sof_dev *sdev, struct snd_sof_widget *
...
+
+	switch (ipc4_copier->dai_type) {
+	case SOF_DAI_INTEL_SSP:
+		switch (clk_type) {
+		case SOF_DAI_CLK_INTEL_SSP_MCLK:
+			return le32_to_cpu(hw_config->mclk_rate);
+		case SOF_DAI_CLK_INTEL_SSP_BCLK:
+			return le32_to_cpu(hw_config->bclk_rate);
+		default:
+			dev_err(sdev->dev, "failed to get clk rate for type: %d\n", clk_type);
```
Comment (ujfalusi) [inline sound/soc/sof/ipc4-topology.c]: "Invalid clk_type for SSP: %d\n" ?

File: sound/soc/sof/ipc4-topology.c [hunk 16]
```diff
@@ -819,6 +837,104 @@ static void sof_ipc4_unprepare_copier_module(struct snd_sof_widget *swidget)
...
+		__func__, dai_index, nhlt_type, dir);
+
+	/* find NHLT blob with matching params */
+	cfg = intel_nhlt_get_endpoint_blob(sdev->dev, ipc4_data->nhlt, dai_index, nhlt_type,
+					   bit_depth, bit_depth, channel_count, sample_rate,
+					   dir, 0);
+
+	if (!cfg) {
+		dev_err(sdev->dev,
+			"no matching blob for sample_rate: %d sample width: %d channels: %d\n",
```
Comment (ujfalusi) [inline sound/soc/sof/ipc4-topology.c]: Drop the `_` from sample_rate

**Review by ujfalusi — CHANGES_REQUESTED** [review]

File: include/uapi/sound/sof/abi.h (hunk 8 above)
Comment (ranj063) [reply include/uapi/sound/sof/abi.h]: @ujfalusi in this particular case, we never send the tlv data to the firmware. does it have to be packed even for reading the data from the topology binary?

Comment (ranj063) [reply include/uapi/sound/sof/abi.h]: I think we should reserve this and unify all the TLV structs in SOF as a follow up patch @ujfalusi. FWIW, when we move to only tuples for IPC data, it will be needed anyway

File: include/uapi/sound/sof/abi.h (hunk 14 above)
Comment (ranj063) [reply include/uapi/sound/sof/abi.h]: are you sure? we only ever get 3 bytes for ABI data in the topology manifest.

Comment (ujfalusi) [reply include/uapi/sound/sof/abi.h]: ipc4 firmware version numbers, not topology. I'm not sure if it is going to cause issues, if we ever compare fw version with topology version.

Comment (ranj063) [reply include/uapi/sound/sof/abi.h]: We add the same version number as the firmware in topology as well. I wonder why it is only 3 bytes. @juimonen any idea?

**Review by RanderWang — APPROVED** [review]
LGTM

File: include/uapi/sound/sof/abi.h (hunk 8 above)
Comment (ujfalusi) [reply include/uapi/sound/sof/abi.h]: Can these be added to the header.h header file where we have `struct sof_abi_hdr` ?

File: include/uapi/sound/sof/abi.h (hunk 14 above)
Comment (ujfalusi) [reply include/uapi/sound/sof/abi.h]: as example fw version: `Booted firmware version: 10.29.0.5399` 5399 is 'build' number in major.minor.hotfix.build

Comment (ranj063) [reply include/uapi/sound/sof/abi.h]: @ujfalusi this is with the cavs FW right? the tplg is built with the SOF repo and it will be of the format major.minor.patch

Comment (ujfalusi) [reply include/uapi/sound/sof/abi.h]: Yes, it was with the cavs FW, but the fields are u16 for both (and SOF also have the dot build in version.

Comment (ranj063) [reply include/uapi/sound/sof/abi.h]: right I think we made a mistake with the ABI in topology by using only 1 byte. fixing it now would mean new tplg will fail with older kernel.

Comment (ranj063) [reply include/uapi/sound/sof/abi.h]: @ujfalusi i've fixed the size of the ABI bytes in manifest now with https://github.com/thesofproject/sof/pull/5631. This will also need an update to @juimonen alsautils PR.

File: sound/soc/sof/intel/hda-dai.c [hunk 17]
```diff
@@ -438,6 +441,88 @@ static int ipc3_hda_dai_trigger(struct snd_pcm_substream *substream,
...
+
+	hstream = substream->runtime->private_data;
+	bus = hstream->bus;
+	rtd = asoc_substream_to_rtd(substream);
+	cpu_dai = asoc_rtd_to_cpu(rtd, 0);
+	codec_dai = asoc_rtd_to_codec(rtd, 0);
+
+	link = snd_hdac_ext_bus_get_link(bus, codec_dai->component->name);
+	if (!link)
+		return -EINVAL;
```
Comment (ujfalusi) [inline sound/soc/sof/intel/hda-dai.c]: I think this would warrant an error print?

File: sound/soc/sof/intel/hda-dai.c [hunk 18]
```diff
@@ -438,6 +441,88 @@ static int ipc3_hda_dai_trigger(struct snd_pcm_substream *substream,
...
+
+		snd_hdac_ext_link_stream_clear(hext_stream);
+		break;
+	}
+	default:
+		return -EINVAL;
+	}
+
+	return 0;
+}
```
Comment (ujfalusi) [inline sound/soc/sof/intel/hda-dai.c:529]: How we will know _why_ `ipc4_hda_link_dma_trigger()` failed from the logs?

File: include/uapi/sound/sof/header.h [hunk 19]
```diff
@@ -26,4 +26,34 @@ struct sof_abi_hdr {
...
+ * @abi_minor: Minor ABI version
+ * @abi_patch: ABI patch
+ * @count: count of tlv items
+ * @items: consecutive variable size tlv items
+ */
+struct sof_manifest {
+	__u16 abi_major;
+	__u16 abi_minor;
+	__u16 abi_patch;
+	__u16 count;
```
Comment (ujfalusi) [inline include/uapi/sound/sof/header.h]: This is coming from the topology file, right? Should not be this of type `__le16`, everything else in generic tplg is LE...

File: sound/soc/sof/ipc3-topology.c [hunk 20]
```diff
@@ -2290,6 +2293,49 @@ static int sof_ipc3_dai_get_clk(struct snd_sof_dev *sdev, struct snd_sof_dai *da
...
+		return 0;
+	}
+
+	if (size != SOF_IPC3_TPLG_ABI_SIZE) {
+		dev_err(scomp->dev, "%s: Invalid topology ABI size\n", __func__);
+		return -EINVAL;
+	}
+
+	dev_info(scomp->dev,
+		 "Topology: ABI %d:%d:%d Kernel ABI %d:%d:%d\n",
```
Comment (ujfalusi) [inline sound/soc/sof/ipc3-topology.c]: I know, copied code, but are we saying that the version is stored as `signed char`?

File: sound/soc/sof/ipc4-topology.c [hunk 21]
```diff
@@ -1315,6 +1316,82 @@ static int sof_ipc4_dai_config(struct snd_sof_dev *sdev, struct snd_sof_widget *
...
+	if (!size || size < SOF_IPC4_TPLG_ABI_SIZE) {
+		dev_err(scomp->dev, "%s: Invalid topology ABI size\n", __func__);
+		return -EINVAL;
+	}
+
+	manifest = (struct sof_manifest *)man_ptr;
+
+	dev_info(scomp->dev,
+		 "Topology: ABI %d:%d:%d Kernel ABI %d:%d:%d\n",
+		  manifest->abi_major, manifest->abi_minor, manifest->abi_patch,
```
Comment (ujfalusi) [inline sound/soc/sof/ipc4-topology.c]: But the manifest->abi_* variables are unsigned?

File: sound/soc/sof/ipc4-topology.c [hunk 22]
```diff
@@ -1315,6 +1316,82 @@ static int sof_ipc4_dai_config(struct snd_sof_dev *sdev, struct snd_sof_widget *
...
+		  SOF_ABI_MAJOR, SOF_ABI_MINOR, SOF_ABI_PATCH);
+
+	abi_version = SOF_ABI_VER(manifest->abi_major, manifest->abi_minor, manifest->abi_patch);
+
+	if (SOF_ABI_VERSION_INCOMPATIBLE(SOF_ABI_VERSION, abi_version)) {
```
Comment (ujfalusi) [inline sound/soc/sof/ipc4-topology.c]: I have a feeling that this and the counterpart in ipc3-topology.c is going to bite us in the future for comparing kernel's defined SOF ABI version to topology ABI version. Just a hunch.

File: sound/soc/sof/ipc4-topology.c [hunk 23]
```diff
@@ -1315,6 +1316,82 @@ static int sof_ipc4_dai_config(struct snd_sof_dev *sdev, struct snd_sof_widget *
...
+				break;
+			ipc4_data->nhlt = devm_kmemdup(sdev->dev, manifest_tlv->data,
+						       manifest_tlv->size, GFP_KERNEL);
+			if (!ipc4_data->nhlt)
+				return -ENOMEM;
+			break;
+		default:
+			dev_err(scomp->dev, "Invalid manifest data type %d\n",
+				manifest_tlv->type);
+			return -EINVAL;
```
Comment (ujfalusi) [inline sound/soc/sof/ipc4-topology.c]: What happens if we just skip the unknown type and print a warn only? Would it be fatal to not handle something we are not aware of?
It could help bisecting since old kernel would not bail out with new firmware which have new type of blob embedded.

**Review by ujfalusi — CHANGES_REQUESTED** [review]

File: include/uapi/sound/sof/header.h (hunk 19 above)
Comment (ranj063) [reply include/uapi/sound/sof/header.h]: you are right. it must be __le16

File: sound/soc/sof/ipc4-topology.c (hunk 23 above)
Comment (ranj063) [reply sound/soc/sof/ipc4-topology.c]: sounds good. ill skip the unknown types.

File: sound/soc/sof/ipc4-topology.c (hunk 22 above)
Comment (ranj063) [reply sound/soc/sof/ipc4-topology.c]: I'm not sure. The idea is that IPC4 topology should only be used when the kernel ABI is updated to ABI 4 as well.

File: sound/soc/sof/intel/hda-dai.c (hunk 17 above)
Comment (ranj063) [reply sound/soc/sof/intel/hda-dai.c]: i 've removed the whol thing. no need assign link at all here.

**Comment by ranj063** [comment]
@juimonen pleae have a look. You'll need to update your alsautils PR to use 2 bytes for ABI major/minor/patch

File: sound/soc/sof/ipc4-topology.c (hunk 22 above)
Comment (ujfalusi) [reply sound/soc/sof/ipc4-topology.c]: Right. SOF_ABI_VERSION_INCOMPATIBLE() is checking the major version, if the kernel's MAJOR is bumped to 4 then old tplg files will not load?

File: sound/soc/sof/ipc4-topology.c (hunk 5 above)
Comment (ujfalusi) [reply sound/soc/sof/ipc4-topology.c:913]: > no, we don't need to set len and dst here

@ranj063, why? Is it guarantied that it is 0 / NULL? If not is there a code somewhere which bases decision on `len` and/or `dst`?

**Review by ujfalusi — DISMISSED** [review]
@ranj063, I think we are going to fall flat on our face with the version check like we have atm, but it is not really this PR which should provide the remedy and I can not say how to handle it right now.

File: sound/soc/sof/ipc4-topology.c (hunk 22 above)
Comment (plbossart) [reply sound/soc/sof/ipc4-topology.c]: I agree with @ujfalusi, the use of ABI checks is just going to hit us big time.
the IPC4 is a totally incompatible solution, it needs to have its own ABI check within the scope of the IPC type. we should NEVER have called this IPC4.

**Comment by ranj063** [comment]
> we need to rethink the ABI checks.

@plbossart @ujfalusi should I drop the ABI check for now and we can revisit this at a later point?

**Comment by ranj063** [comment]
> > we need to rethink the ABI checks.
> 
> @plbossart @ujfalusi should I drop the ABI check for now and we can revisit this at a later point?

I've replaced the check with a todo now.

File: sound/soc/sof/ipc4-topology.c [hunk 24]
```diff
@@ -516,7 +533,8 @@ static void sof_ipc4_widget_free_comp_dai(struct snd_sof_widget *swidget)
 	kfree(available_fmt->dma_buffer_size);
 	kfree(available_fmt->base_config);
 	kfree(available_fmt->out_audio_fmt);
-	kfree(ipc4_copier->copier_config);
+	if (ipc4_copier->dai_type == SOF_DAI_INTEL_HDA)
+		kfree(ipc4_copier->copier_config);
```
Comment (bardliao) [inline sound/soc/sof/ipc4-topology.c:537]: Should this be `if (ipc4_copier->dai_type != SOF_DAI_INTEL_SSP && ipc4_copier->dai_type != SOF_DAI_INTEL_DMIC)`? We allocate `ipc4_copier->copier_config` in the default case in `sof_ipc4_widget_setup_comp_dai`, so I think we should free `ipc4_copier->copier_config` in all dai_type except `SOF_DAI_INTEL_SSP` and `SOF_DAI_INTEL_DMIC`.
We will allocate ipc4_copier->copier_config in `SOF_DAI_INTEL_ALH` type, too.

Comment (bardliao) [reply sound/soc/sof/ipc4-topology.c:537]: > Should this be `if (ipc4_copier->dai_type != SOF_DAI_INTEL_SSP && ipc4_copier->dai_type != SOF_DAI_INTEL_DMIC)`? We allocate `ipc4_copier->copier_config` in the default case in `sof_ipc4_widget_setup_comp_dai`, so I think we should free `ipc4_copier->copier_config` in all dai_type except `SOF_DAI_INTEL_SSP` and `SOF_DAI_INTEL_DMIC`. We will allocate ipc4_copier->copier_config in `SOF_DAI_INTEL_ALH` type, too.

Chat with @ranj063, I will merge it now and fix it with a fixup commit,

**Review by bardliao — APPROVED** [review]
